other/web-js/client/src/client-api.js

```diff
@@ -0,0 +1,312 @@
+    }
+    return `https://${region}.${cloud}.realm.mongodb.com/api/client/v2.0/`;
+  }
+
+  /**
+   * Create a new Client API client.
+   * @param {object} config The configuration for the Client API client.
+   * @param {string} config.appId The Client App ID of the App Services App to connect to, e.g. `myapp-abcde`.
+   * @param {string} [config.deployment_model] The deployment model of the App Services App to connect to. Defaults to `GLOBAL` if `cloud` and `region` are not specified.
+   * @param {string} [config.cloud] The cloud provider to connect to. Required if `region` is specified.
+   * @param {string} [config.region] The region to connect to. Required if `cloud` is specified.
+   * @param {function} [config.onAuthChange] A callback that's run with the latest auth state whenever the current user changes.
+   * @example
+   * const clientApi = new ClientApi({
+   *   appId: "myapp-abcde",
+   *   deployment_model: "LOCAL",
+   *   cloud: "aws",
+   *   region: "us-east-1",
+   *    onAuthChange: (currentUser) => {
+   *     console.log("The current user is now:", currentUser.id);
+   *   }
+   * });
+   */
+  constructor({ appId, deployment_model, cloud, region, onAuthChange }) {
+    this.appId = appId;
+    this.credentialStorage = new CredentialStorage(appId);
+    this.currentUser = this.credentialStorage.get("currentUser");
+    this.deployment = {
+      deployment_model:
+        deployment_model ?? (cloud || region) ? "LOCAL" : "GLOBAL",
+      cloud,
+      region,
+    };
+    this.baseUrl = ClientApi.constructBaseUrl(this.deployment);
+    this.onAuthChange = onAuthChange;
+  }
+
+  /**
+   * Construct a Client API URL at the given path
+   * @param {string} path The path to the Client API endpoint, e.g. `/auth/providers/local-userpass/login`.
+   * @returns {string} The full URL of the Client API endpoint.
+   */
+  #endpointUrl = (path) => {
+    if (!path.startsWith("/")) {
+      throw new Error(`Client API path must start with a slash ("/")`);
+    }
+    const url = new URL(`app/${this.appId}` + path, this.baseUrl);
+    return url.href;
+  };
+
+  /**
+   * @typedef {object} User User account information, session tokens, and other metadata.
+   * @property {string} user_id The user's account ID.
+   * @property {string} access_token The access token for the user.
+   * @property {string} refresh_token The refresh token for the user.
+   * @property {string} [device_id] The ID of the device that the user is logged in on.
+   */
+
+  /**
+   * Set the current user.
+   * @param {User | null} user The user to set as the current user, or null to unauthenticate the client.
+   * @example
+   * this.#setCurrentUser({
+   *   "user_id": "63f504a25dde21e6f10ba025"
+   *   "access_token": "eyJhbGciOiJIUzI1NiIsI...TYNZYyCM3RKl77FuW_kBf29POr24"
+   *   "device_id": "63f504a75dd881f2f58ba09a",
+   *   "refresh_token": "eyJhbGciOiJIUzI1BiIsI...ynIqiC0bxVAdvYJmw3L1358T3UJk",
+   * })
+   */
+  #setCurrentUser = (user) => {
```

Comment:
this comment applies to all the methods on this class. i notice that your methods are all arrow functions (as opposed to `function(){}` declarations. and also these methods make frequent use of the `this` keyword. 

arrow functions can have a funky relationship with `this` when used in JS classes (see [this article](https://www.codementor.io/@dariogarciamoya/understanding-this-in-javascript-with-arrow-functions-gcpjwfyuc#:~:text=In%20my%20opinion%2C%20arrow%20functions%20are%20the%20best%20choice%20when%20working%20with%20closures%20or%20callbacks%2C%20but%20not%20a%20good%20choice%20when%20working%20with%20class/object%20methods%20or%20constructors.)). 

i haven't looked thru your code to make sure that none of this 'funk' is creeping in. but in order to mitigate against this happening either in the template app or a user adopting the template app consider replacing all arrow function methods with 'normal' function declaration methods.

Reply:
The binding inference is the point here - otherwise you need to call a bunch of `.bind(this)` in the hooks